packages/smart-wallet/src/offerWatcher.js

```diff
@@ -0,0 +1,227 @@
+/* global globalThis */
+
+import { E, passStyleOf } from '@endo/far';
+
+import { isUpgradeDisconnection } from '@agoric/internal/src/upgrade-api.js';
+import { prepareExoClassKit } from '@agoric/vat-data';
+import { M } from '@agoric/store';
+import {
+  PaymentPKeywordRecordShape,
+  SeatShape,
+} from '@agoric/zoe/src/typeGuards.js';
+import { AmountShape } from '@agoric/store/test/borrow-guards.js';
+import { objectMap, deeplyFulfilledObject } from '@agoric/internal';
+
+/** Value for "result" field when the result can't be published */
+export const UNPUBLISHED_RESULT = 'UNPUBLISHED';
+
+/**
+ * @typedef {import('./smartWallet.js').OfferSpec & {
+ * error?: string,
+ * numWantsSatisfied?: number
+ * result?: unknown | typeof UNPUBLISHED_RESULT,
+ * payouts?: AmountKeywordRecord,
+ * }} OfferStatus
+ */
+
+const watchForOfferResult = (watcher, seat) => {
+  const rWatcher = watcher.resultWatcher;
+  globalThis.VatData.watchPromise(E(seat).getOfferResult(), rWatcher, seat);
+};
+
+const watchForNumWants = (watcher, seat) => {
+  const nWatcher = watcher.numWantsWatcher;
+  globalThis.VatData.watchPromise(E(seat).numWantsSatisfied(), nWatcher, seat);
+};
+
+const watchForPayout = (watcher, seat) => {
+  const pWatcher = watcher.paymentWatcher;
+  globalThis.VatData.watchPromise(E(seat).getPayouts(), pWatcher, seat);
+};
+
+export const watchOfferOutcomes = (watcher, seat) => {
+  watchForPayout(watcher, seat);
+  watchForNumWants(watcher, seat);
+  watchForOfferResult(watcher, seat);
```

Comment:
the helper methods maybe obscure more than clarify. consider inlining

```suggestion
export const watchOfferOutcomes = (watcher, seat) => {
  watchPromise(E(seat).getOfferResult(), watcher.resultWatcher, seat);
  watchPromise(E(seat).numWantsSatisfied(), watcher.numWantsWatcher, seat);
  watchPromise(E(seat).getPayouts(), watcher.paymentWatcher, seat);
};
```

Reply:
also please annotate the params

Reply:
The helper methods area also called in the watchers on disconnections.

Reply:
I tried declaring `watcher` to be ` {ReturnType<MakeOfferWatcher>}`, but tsc says 
```
src/offerWatcher.js:195:28 - error TS2345: Argument of type '{ helper: { updateStatus(offerStatusUpdates: any): void; onNewContinuingOffer(offerId: any, invitationAmount: any, invitationMakers: any, publicSubscribers: any): void; publishResult(result: any): void; exitOpenSeat(reason: any): never; }; paymentWatcher: { ...; }; resultWatcher: { ...; }; numWantsWatcher: { ...; }; }' is not assignable to parameter of type '{ helper: { updateStatus(offerStatusUpdates: any): void; onNewContinuingOffer(offerId: any, invitationAmount: any, invitationMakers: any, publicSubscribers: any): void; publishResult(result: any): void; exitOpenSeat(reason: any): never; }; paymentWatcher: { ...; }; resultWatcher: { ...; }; numWantsWatcher: { ...; };...'.
  Type '{ helper: { updateStatus(offerStatusUpdates: any): void; onNewContinuingOffer(offerId: any, invitationAmount: any, invitationMakers: any, publicSubscribers: any): void; publishResult(result: any): void; exitOpenSeat(reason: any): never; }; paymentWatcher: { ...; }; resultWatcher: { ...; }; numWantsWatcher: { ...; }; }' is missing the following properties from type 'RemotableBrand<{}, { helper: { updateStatus(offerStatusUpdates: any): void; onNewContinuingOffer(offerId: any, invitationAmount: any, invitationMakers: any, publicSubscribers: any): void; publishResult(result: any): void; exitOpenSeat(reason: any): never; }; paymentWatcher: { ...; }; resultWatcher: { ...; }; numWant...': L, R
```
which I don't understand.